src/classlibnative/bcltype/bignum.h

```diff
@@ -0,0 +1,140 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: bignum.h
+//
+
+//
+
+#ifndef _BIGNUM_H_
+#define _BIGNUM_H_
+
+#include <clrtypes.h>
+
+class BigNum
+{
+public:
+    BigNum();
+    BigNum(UINT32 value);
+    BigNum(UINT64 value);
+    ~BigNum();
+
+    BigNum & operator=(const BigNum &rhs);
+
+    static UINT32 LogBase2(UINT32 val);
+    static UINT32 LogBase2(UINT64 val);
+
+    static int Compare(const BigNum& lhs, UINT32 value);
+    static int Compare(const BigNum& lhs, const BigNum& rhs);
+
+    static void ShiftLeft(UINT64 input, int shift, BigNum& output);
+    static void ShiftLeft(BigNum* pResult, UINT32 shift);
+    static void Pow10(int exp, BigNum& result);
+    static void PrepareHeuristicDivide(BigNum* pDividend, BigNum* divisor);
+    static UINT32 HeuristicDivide(BigNum* pDividend, const BigNum& divisor);
+    static void Multiply(const BigNum& lhs, UINT32 value, BigNum& result);
+    static void Multiply(const BigNum& lhs, const BigNum& rhs, BigNum& result);
+
+    bool IsZero();
+
+    void Multiply(UINT32 value);
+    void Multiply(const BigNum& value);
+    void SetUInt32(UINT32 value);
+    void SetUInt64(UINT64 value);
+    void ExtendBlock(UINT32 newBlock);
+
+private:
+
+    static const UINT8 BIGSIZE = 35;
+    static const UINT8 UINT32POWER10NUM = 8;
+    static const UINT8 BIGPOWER10NUM = 6;
+    static UINT32 m_power10UInt32Table[UINT32POWER10NUM];
+    static BigNum m_power10BigNumTable[BIGPOWER10NUM];
+
+    static class StaticInitializer
+    {
+    public:
+        StaticInitializer()
+        {
+            // 10^8
```

Comment:
This is a little bit tricky. It's along with m_power10UInt32Table to speed up BigNum::Pow10(exp, result). If we have a more efficient implementation I can remove this one.

Let's say we pass a value whose binary representation is `111111` to `BigNum::Pow10(exp, result)`.

We can split the exp to two parts, one is small exp, which 10^smallExp can be represented as UINT32, another part is 10^bigExp, which must be represented as BigNum. So the result should be `10^smallExp * 10^bigExp`.

Calculate 10^smallExp is simple, we just lookup the 10^smallExp from `m_power10UInt32Table`. But here's a bad news: Although UINT32 can represent 10^9, exp 9's binary representation is `1001`. That means 10^(1011), 10^(1101), 10^(1111) all cannot be stored as UINT32, we cannot easily say something like: "If exp <= 3 bits, its a smallExp, if exp > 3 bits, its a bigExp". So instead of involving 10^8, 10^9 to m_power10UInt32Table, I consider 10^8, 10^9 as a bigNum, so they fall into `m_power10BigNumTable`. Now we can have a simple rule: "If exp <= 3 bits, its a smallExp, if exp > 3 bits, its a bigExp".

For (111111)2, we first calculate 10^(smallExp), which is 10^(7), now we can shift right 3 bits, prepare to calculate the bigExp part, the exp now becomes (000111)2. 

Apparently the first bit of bigExp should reperesent 10^8 because we have already shifted 3 bits for smallExp, so m_power10BigNumTable[0] = 10^8.
Now let's shift exp right 1 more bit, the second bigExp bit should represent 10^(8 * 2) = 10^16. Then so on...

That's why we just need the value of m_power10BigNumTable be power of 2.

This idea is from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/ . I think it's tricky but faster than a raw pow10 implementation. I'm still looking for a more efficient and elegant pow10 implementation. If there is, I'd like to replace this one.